pkg/tempdir/tempdir.go

```diff
@@ -0,0 +1,259 @@
+    (.lock file inside it). This part of the process does *not* hold the 'tempDirGlobalLock'.
+  - If TryLock() on the 'instanceLock' succeeds: The directory is considered stale. The 'instanceLock' is
+    immediately Unlock()ed, and then os.RemoveAll() is called on the directory.
+    Unlocking before removal is crucial to:
+    1. Prevent holding the 'instanceLock' during a potentially slow os.RemoveAll.
+    2. Allow os.RemoveAll to operate on an unlocked directory.
+    3. If os.RemoveAll fails, the directory remains unlocked for future recovery attempts.
+  - If TryLock() on the 'instanceLock' fails: The directory is considered in active use by another
+    process/goroutine (which holds its 'instanceLock'), and it's skipped.
+
+4. TempDir Usage:
+  - TempDir.Add() requires the 'instanceLock' (td.instanceLock) to be held (enforced by AssertLocked).
+
+5. Cleanup Timing (Caller's Responsibility for Application-Level Lock):
+  - TempDir.Cleanup() itself acquires the 'tempDirGlobalLock' for its duration.
+  - Callers SHOULD defer TempDir.Cleanup() to run *after* their broader
+    application-level global lock (see point 1) has been released.
+    This prevents potentially slow disk I/O during Cleanup() from blocking
+    other critical application operations that might be waiting on that
+    application-level lock.
+
+Example structure:
+
+	rootDir/
+	  .global-tmp-lock  (tempDirGlobalLock)
+	  temp-dir-ABC/
+	    .lock           (instanceLock for temp-dir-ABC)
+	    file1
+	    file3
+	  temp-dir-XYZ/
+	    .lock           (instanceLock for temp-dir-XYZ)
+	    file2
+*/
+const (
+	// TempDirPrefix is the prefix used for creating temporary directories.
+	TempDirPrefix = "temp-dir-"
+	// globalLockFileName is the name of the lock file in the root directory.
+	globalLockFileName = ".global-tmp-lock"
+	// instanceLockFileName is the name of the lock file within each temporary directory.
+	instanceLockFileName = ".lock"
+)
+
+// TempDir represents a temporary directory that is created in a specified root directory.
+type TempDir struct {
+	rootDir string
+	// tempDirGlobalLock is a lock file (e.g., rootDir/.global-tmp-lock) used to ensure
+	// that listing for recovery, creation, and cleanup of temporary directories for this
+	// rootDir are serialized across processes and goroutines.
+	tempDirGlobalLock *lockfile.LockFile
+
+	tempDirPath string
+	// instanceLock is a lock file (e.g., rootDir/temp-dir-XYZ/.lock) specific to this
+	// TempDir instance, indicating it's in active use.
+	instanceLock *lockfile.LockFile
+}
+
+// listPotentialStaleDirs scans the rootDir for directories that might be stale temporary directories.
+// It acquires and releases the tempDirGlobalLock to ensure safe concurrent access to the directory listing.
+func (td *TempDir) listPotentialStaleDirs(rootDirPath string) ([]string, error) {
+	td.tempDirGlobalLock.Lock()
+	defer td.tempDirGlobalLock.Unlock()
```

Comment:
we don't need the lock here since it is released when the function exits, so for the caller it doesn't matter whether the code below runs with the lock or not.

As @mtrmac pointed out, we need to use `os.ReadDir` not Walk it recursively